File: stripe/src/main/java/com/stripe/android/view/ExpiryDateEditText.java
package com.stripe.android.view;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.Size;
import android.support.annotation.VisibleForTesting;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.widget.EditText;

/**
 * An {@link EditText} that handles putting numbers around a central divider character.
 */
public class ExpiryDateEditText extends StripeEditText {

    static final int INVALID_INPUT = -1;
    private static final int MAX_INPUT_LENGTH = 5;

    private ExpiryDateEditListener mExpiryDateEditListener;
    private boolean mIsDateValid;

    public ExpiryDateEditText(Context context) {
        super(context);
        listenForTextChanges();
    }

    public ExpiryDateEditText(Context context, AttributeSet attrs) {
        super(context, attrs);
        listenForTextChanges();
    }

    public ExpiryDateEditText(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        listenForTextChanges();
    }

    /**
     * Gets whether or not the date currently entered is valid and not yet
     * passed.
     *
     * @return {@code true} if the text entered represents a valid expiry date that has not
     * yet passed, and {@code false} if not.
     */
    public boolean isDateValid() {
        return mIsDateValid;
    }

    /**
     * Gets the expiry date displayed on this control if it is valid, or {@code null} if it is not.
     * The return value is given as a length-2 {@code int} array, where the first entry is the
     * two-digit month (from 01-12) and the second entry is the four-digit year (2017, not 17).
     *
     * @return an {@code int} array of the form {month, year} if the date is valid, or {@code null}
     * if it is not
     */
    @Nullable
    @Size(2)
    public int[] getValidDateFields() {
        if (!mIsDateValid) {
            return null;
        }

        int [] monthYearPair = new int[2];
        String rawNumericInput = getText().toString().replaceAll("/", "");
        String[] dateFields = DateUtils.separateDateStringParts(rawNumericInput);

        try {
            monthYearPair[0] = Integer.parseInt(dateFields[0]);
            monthYearPair[1] = DateUtils.convertTwoDigitYearToFour(Integer.parseInt(dateFields[1]));
        } catch (NumberFormatException numEx) {
            // Given that the date should already be valid when getting to this method, we should
            // not his this exception. Returning null to indicate error if we do.
            return null;
        }

        return monthYearPair;
    }

    public void setExpiryDateEditListener(ExpiryDateEditListener expiryDateEditListener) {
        mExpiryDateEditListener = expiryDateEditListener;
    }

    private void listenForTextChanges() {
        addTextChangedListener(new TextWatcher() {
            boolean ignoreChanges = false;
            int latestChangeStart;
            int latestInsertionSize;
            String[] parts = new String[2];

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                if (ignoreChanges) {
                    return;
                }
                latestChangeStart = start;
                latestInsertionSize = after;
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (ignoreChanges) {
                    return;
                }

                boolean inErrorState = false;

                String rawNumericInput = s.toString().replaceAll("/", "");

                if (rawNumericInput.length() == 1
                        && latestChangeStart == 0
                        && latestInsertionSize == 1) {
                    char first = rawNumericInput.charAt(0);
                    if (!(first == '0' || first == '1')) {
                        // If the first digit typed isn't 0 or 1, then it can't be a valid
                        // two-digit month. Hence, we assume the user is inputting a one-digit
                        // month. We bump it to the preferred input, so "4" becomes "04", which
                        // later in this method goes to "04/".
                        rawNumericInput = "0" + rawNumericInput;
                        latestInsertionSize++;
                    }
                } else if (rawNumericInput.length() == 2
                        && latestChangeStart == 2
                        && latestInsertionSize == 0) {
                    // This allows us to delete past the separator, so that if a user presses
                    // delete when the current string is "12/", the resulting string is "1," since
                    // we pretend that the "/" isn't really there. The case that we also want,
                    // where "12/3" + DEL => "12" is handled elsewhere.
                    rawNumericInput = rawNumericInput.substring(0, 1);
                }

                // Date input is MM/YY, so the separated parts will be {MM, YY}
                parts = DateUtils.separateDateStringParts(rawNumericInput);

                if (!DateUtils.isValidMonth(parts[0])) {
                    inErrorState = true;
                }

                StringBuilder formattedDateBuilder = new StringBuilder();
                formattedDateBuilder.append(parts[0]);
                // parts[0] is the two-digit month
                if ((parts[0].length() == 2 && latestInsertionSize > 0 && !inErrorState)
                        || rawNumericInput.length() > 2) {
                    formattedDateBuilder.append("/");
                }
                // parts[1] is the two-digit year
                formattedDateBuilder.append(parts[1]);

                String formattedDate = formattedDateBuilder.toString();
                int cursorPosition = updateSelectionIndex(
                        formattedDate.length(),
                        latestChangeStart,
                        latestInsertionSize,
                        MAX_INPUT_LENGTH);

                ignoreChanges = true;
                setText(formattedDate);
                setSelection(cursorPosition);
                ignoreChanges = false;
            }

            @Override
            public void afterTextChanged(Editable s) {
                // Note: we want to show an error state if the month is invalid or the
                // final, complete date is in the past. We don't want to show an error state for
                // incomplete entries.
                boolean shouldShowError = false;
                if (parts[0].length() == 2 && !DateUtils.isValidMonth(parts[0])) {
                    // This covers the case where the user has entered a month of 15, for instance.
                    shouldShowError = true;
                }

                // Note that we have to check the parts array because afterTextChanged has odd
                // behavior when it comes to pasting, where a paste of "1212" triggers this
                // function for the strings "12/12" (what it actually becomes) and "1212",
                // so we might not be properly catching an error state.
                if (parts[0].length() == 2 && parts[1].length() == 2) {
                    boolean wasComplete = mIsDateValid;
                    updateInputValues(parts);
                    // Here, we have a complete date, so if we've made an invalid one, we want
                    // to show an error.
                    shouldShowError = !mIsDateValid;
                    if (!wasComplete && mIsDateValid && mExpiryDateEditListener != null) {
                        mExpiryDateEditListener.onExpiryDateComplete();
                    }
                } else {
                    mIsDateValid = false;
                }

                setShouldShowError(shouldShowError);
            }
        });
    }

    /**
     * Updates the selection index based on the current (pre-edit) index, and
     * the size change of the number being input.
     *
     * @param newLength the post-edit length of the string
     * @param editActionStart the position in the string at which the edit action starts
     * @param editActionAddition the number of new characters going into the string (zero for
     *                           delete)
     * @return an index within the string at which to put the cursor
     */
    @VisibleForTesting
    int updateSelectionIndex(
            int newLength,
            int editActionStart,
            int editActionAddition,
            int maxInputLength) { int newPosition, gapsJumped = 0;

        boolean skipBack = false;
        if (editActionStart <= 2 && editActionStart + editActionAddition >= 2) {
            gapsJumped = 1;
        }

        // editActionAddition can only be 0 if we are deleting,
        // so we need to check whether or not to skip backwards one space
        if (editActionAddition == 0 && editActionStart == 3) {
            skipBack = true;
        }

        newPosition = editActionStart + editActionAddition + gapsJumped;
        if (skipBack && newPosition > 0) {
            newPosition--;
        }
        int untruncatedPosition = newPosition <= newLength ? newPosition : newLength;
        return Math.min(maxInputLength, untruncatedPosition);
    }

    private void updateInputValues(@NonNull @Size(2) String[] parts) {
        int inputMonth;
        int inputYear;
        if (parts[0].length() != 2) {
            inputMonth = INVALID_INPUT;
        } else {
            try {
                inputMonth = Integer.parseInt(parts[0]);
            } catch (NumberFormatException numEx) {
                inputMonth = INVALID_INPUT;
            }
        }

        if (parts[1].length() != 2) {
            inputYear = INVALID_INPUT;
        } else {
            try {
                inputYear = DateUtils.convertTwoDigitYearToFour(Integer.parseInt(parts[1]));
            } catch (NumberFormatException numEx) {
                inputYear = INVALID_INPUT;
            }
        }

        mIsDateValid = DateUtils.isExpiryDataValid(inputMonth, inputYear);
    }

    interface ExpiryDateEditListener {
        void onExpiryDateComplete();
    }
}

File: stripe/src/main/java/com/stripe/android/view/ExpiryDateEditText.kt
package com.stripe.android.view

import android.content.Context
import android.os.Build
import android.text.Editable
import android.text.InputFilter
import android.text.InputType
import android.util.AttributeSet
import android.view.View
import android.widget.EditText
import androidx.annotation.VisibleForTesting
import com.stripe.android.R
import com.stripe.android.model.ExpirationDate
import kotlin.math.min

/**
 * An [EditText] that handles putting numbers around a central divider character.
 */
class ExpiryDateEditText @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = androidx.appcompat.R.attr.editTextStyle
) : StripeEditText(context, attrs, defStyleAttr) {

    init {
        inputType = InputType.TYPE_CLASS_NUMBER
        filters = listOf(
            InputFilter.LengthFilter(
                context.resources.getInteger(R.integer.stripe_date_length)
            )
        ).toTypedArray()

        setErrorMessage(resources.getString(R.string.invalid_expiry_year))
        listenForTextChanges()

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            setAutofillHints(View.AUTOFILL_HINT_CREDIT_CARD_EXPIRATION_DATE)
        }
    }

    // invoked when a valid date has been entered
    @JvmSynthetic
    internal var completionCallback: () -> Unit = {}

    /**
     * Is `true` if the text entered represents a valid expiry date that has not
     * yet passed, and `false` if not.
     */
    var isDateValid: Boolean = false
        private set

    /**
     * Gets the expiry date displayed on this control if it is valid, or `null` if it is not.
     * The return value is given as a [Pair], where the first entry is the two-digit month
     * (from 01-12) and the second entry is the four-digit year (2017, not 17).
     */
    @Deprecated("Use validatedDate")
    val validDateFields: Pair<Int, Int>?
        get() = validatedDate?.let {
            Pair(it.month, it.year)
        }

    val validatedDate: ExpirationDate.Validated?
        get() {
            return when (isDateValid) {
                true -> ExpirationDate.Unvalidated.create(fieldText).validate()
                false -> null
            }
        }

    override val accessibilityText: String?
        get() {
            return resources.getString(R.string.acc_label_expiry_date_node, text)
        }

    private fun listenForTextChanges() {
        addTextChangedListener(
            object : StripeTextWatcher() {
                private var ignoreChanges = false
                private var latestChangeStart: Int = 0
                private var latestInsertionSize: Int = 0
                private var expirationDate = ExpirationDate.Unvalidated(
                    month = "",
                    year = ""
                )

                private var newCursorPosition: Int? = null
                private var formattedDate: String? = null

                override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
                    if (ignoreChanges) {
                        return
                    }
                    latestChangeStart = start
                    latestInsertionSize = after
                }

                override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                    if (ignoreChanges) {
                        return
                    }

                    var inErrorState = false

                    val inputText = s?.toString().orEmpty()
                    var rawNumericInput = inputText.replace(SEPARATOR.toRegex(), "")

                    if (rawNumericInput.length == 1 && latestChangeStart == 0 &&
                        latestInsertionSize == 1
                    ) {
                        val first = rawNumericInput[0]
                        if (!(first == '0' || first == '1')) {
                            // If the first digit typed isn't 0 or 1, then it can't be a valid
                            // two-digit month. Hence, we assume the user is inputting a one-digit
                            // month. We bump it to the preferred input, so "4" becomes "04", which
                            // later in this method goes to "04/".
                            rawNumericInput = "0$rawNumericInput"
                            latestInsertionSize++
                        }
                    } else if (rawNumericInput.length == 2 &&
                        latestChangeStart == 2 &&
                        latestInsertionSize == 0
                    ) {
                        // This allows us to delete past the separator, so that if a user presses
                        // delete when the current string is "12/", the resulting string is "1," since
                        // we pretend that the "/" isn't really there. The case that we also want,
                        // where "12/3" + DEL => "12" is handled elsewhere.
                        rawNumericInput = rawNumericInput.substring(0, 1)
                    }

                    val expirationDate =
                        ExpirationDate.Unvalidated.create(rawNumericInput).also {
                            this.expirationDate = it
                        }

                    if (!expirationDate.isMonthValid) {
                        inErrorState = true
                    }

                    val formattedDateBuilder = StringBuilder()
                        .append(expirationDate.month)

                    if (expirationDate.month.length == 2 && latestInsertionSize > 0 &&
                        !inErrorState || rawNumericInput.length > 2
                    ) {
                        formattedDateBuilder.append(SEPARATOR)
                    }

                    formattedDateBuilder.append(expirationDate.year)

                    val formattedDate = formattedDateBuilder.toString()
                    this.newCursorPosition = updateSelectionIndex(
                        formattedDate.length,
                        latestChangeStart,
                        latestInsertionSize,
                        MAX_INPUT_LENGTH
                    )
                    this.formattedDate = formattedDate
                }

                override fun afterTextChanged(s: Editable?) {
                    if (ignoreChanges) {
                        return
                    }

                    ignoreChanges = true
                    if (!isLastKeyDelete && formattedDate != null) {
                        setText(formattedDate)
                        newCursorPosition?.let {
                            setSelection(it.coerceIn(0, fieldText.length))
                        }
                    }

                    ignoreChanges = false

                    // Note: we want to show an error state if the month is invalid or the
                    // final, complete date is in the past. We don't want to show an error state for
                    // incomplete entries.

                    // This covers the case where the user has entered a month of 15, for instance.
                    val month = expirationDate.month
                    val year = expirationDate.year
                    var shouldShowError = month.length == 2 && !expirationDate.isMonthValid

                    // Note that we have to check the parts array because afterTextChanged has odd
                    // behavior when it comes to pasting, where a paste of "1212" triggers this
                    // function for the strings "12/12" (what it actually becomes) and "1212",
                    // so we might not be properly catching an error state.
                    if (month.length == 2 && year.length == 2) {
                        val wasComplete = isDateValid
                        isDateValid = isDateValid(month, year)
                        // Here, we have a complete date, so if we've made an invalid one, we want
                        // to show an error.
                        shouldShowError = !isDateValid
                        if (!wasComplete && isDateValid) {
                            completionCallback()
                        }
                    } else {
                        isDateValid = false
                    }

                    this@ExpiryDateEditText.shouldShowError = shouldShowError

                    formattedDate = null
                    newCursorPosition = null
                }
            }
        )
    }

    /**
     * Updates the selection index based on the current (pre-edit) index, and
     * the size change of the number being input.
     *
     * @param newLength the post-edit length of the string
     * @param editActionStart the position in the string at which the edit action starts
     * @param editActionAddition the number of new characters going into the string (zero for
     * delete)
     *
     * @return an index within the string at which to put the cursor
     */
    @VisibleForTesting
    internal fun updateSelectionIndex(
        newLength: Int,
        editActionStart: Int,
        editActionAddition: Int,
        maxInputLength: Int
    ): Int {
        val gapsJumped = if (editActionStart <= 2 && editActionStart + editActionAddition >= 2) {
            SEPARATOR.length
        } else {
            0
        }

        // `shouldRemoveSeparator` will be true when deleting the character immediately after the
        // separator. For example, if the input text is currently "01/2" and a character is
        // deleted, both the '2' and the separator should be deleted.
        val isDelete = editActionAddition == 0
        val shouldRemoveSeparator = isDelete && editActionStart == (2 + SEPARATOR.length)

        val newPosition = (editActionStart + editActionAddition + gapsJumped).let { newPosition ->
            newPosition - if (shouldRemoveSeparator && newPosition > 0) {
                SEPARATOR.length
            } else {
                0
            }
        }

        val untruncatedPosition = min(newPosition, newLength)
        return min(maxInputLength, untruncatedPosition)
    }

    private fun isDateValid(month: String, year: String): Boolean {
        val inputMonth = if (month.length != 2) {
            INVALID_INPUT
        } else {
            runCatching {
                month.toInt()
            }.getOrDefault(INVALID_INPUT)
        }

        val inputYear = if (year.length != 2) {
            INVALID_INPUT
        } else {
            runCatching {
                DateUtils.convertTwoDigitYearToFour(year.toInt())
            }.getOrDefault(INVALID_INPUT)
        }

        return DateUtils.isExpiryDataValid(inputMonth, inputYear)
    }

    private companion object {
        private const val SEPARATOR = "/"
        private const val INVALID_INPUT = -1
        private const val MAX_INPUT_LENGTH = 4 + SEPARATOR.length
    }
}
